### ISSUE TITLE
[Issue]: Assertion `Index < Length && "Invalid index!"' failed

### ISSUE THREAD

**Issue by causten** [ROCm/rocMLIR#1417]
### Problem Description

Build MIGraphX `develop` branch + https://github.com/ROCm/AMDMIGraphX/pull/2782 
I was using a ROCm 6.0.2 docker
I ran the migraphx-driver and hit the error listed in the "Additional Information" section of this issue.

### Operating System

20.04.6

### CPU

AMD EPYC 7702 64-Core Processor

### GPU

AMD Instinct MI100

### ROCm Version

ROCm 6.0.0

### ROCm Component

rocMLIR

### Additional Information

Trace collected from MIGRAPHX_TRACE_MLIR=1 /workspace/AMDMIGraphX/build/bin/driver perf /new-saved-models/tiny_llama/decoder_model.onnx --fill1 input_ids attention_mask --inpu
t-dim @input_ids 1 256 @attention_mask 1 256

[mlir_error_tinylama.txt](https://github.com/ROCm/rocMLIR/files/14316147/mlir_error_tinylama.txt)

**Comment by krzysz00** [comment]
Having isolated out failing examples out of the parallel compilation (I think we need a future note to `MIGRAPHX_GPU_COMPILE_PARALLEL=1` ... that variable really needs to be renamed ... when collecting traces), I've got the following example

```mlir
module {
  func.func @mlir_reshape_reshape_transpose_dot(%arg0: !migraphx.shaped<1x32x256x64xf32, 524288x64x2048x1>, %arg1: !migraphx.shaped<1x4x256x64xf32, 65536x16384x64x1>) -> !migraphx.shaped<1x32x256x256xf32, 2097152x65536x256x1> attributes {arch = "gfx908:sramecc+:xnack-", kernel = "mixr", num_cu = 120 : i64} {
    %0 = migraphx.reshape %arg1 {dims = [1, 4, 1, 256, 64]} : <1x4x256x64xf32, 65536x16384x64x1> -> <1x4x1x256x64xf32, 65536x16384x16384x64x1>
    %1 = migraphx.multibroadcast %0 {out_dyn_dims = [], out_lens = [1, 4, 8, 256, 64]} : <1x4x1x256x64xf32, 65536x16384x16384x64x1> -> <1x4x8x256x64xf32, 65536x16384x0x64x1>
    %2 = migraphx.reshape %1 {dims = [1, 32, 256, 64]} : <1x4x8x256x64xf32, 65536x16384x0x64x1> -> <1x32x256x64xf32, 524288x16384x64x1>
    %3 = migraphx.transpose %2 {permutation = [0, 1, 3, 2]} : <1x32x256x64xf32, 524288x16384x64x1> -> <1x32x64x256xf32, 524288x16384x1x64>
    %4 = migraphx.dot %arg0, %3 : <1x32x256x64xf32, 524288x64x2048x1>, <1x32x64x256xf32, 524288x16384x1x64> -> <1x32x256x256xf32, 2097152x65536x256x1>
    return %4 : !migraphx.shaped<1x32x256x256xf32, 2097152x65536x256x1>
  }
}
```

which crashes with the following backtrace
```
% ./bin/rocmlir-driver --kernel-pipeline=migraphx,highlevel ./1417-ex1.mlir
rocmlir-driver: /home/kdrewnia/rocmlir/external/llvm-project/llvm/include/llvm/ADT/ArrayRef.h:255: const T &llvm::ArrayRef<long>::operator[](size_t) const [T = long]: Assertion `Index < Length && "Invalid index!"' failed. 
...
#11 0x00007f4fc6715233 (anonymous namespace)::FoldBroadcast::isBatchDimFoldableInTheTransformStack(mlir::ArrayAttr) const /home/kdrewnia/rocmlir/mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:0:0
#12 0x00007f4fc6715233 (anonymous namespace)::FoldBroadcast::isBatchDimFoldable(mlir::PatternRewriter&, mlir::Value) const /home/kdrewnia/rocmlir/mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:131:12
#13 0x00007f4fc671481a (anonymous namespace)::FoldBroadcast::matchAndRewrite(mlir::rock::GemmOp, mlir::PatternRewriter&) const /home/kdrewnia/rocmlir/mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp:173:30      
....
```

I'll be investigating further.

So far, this finding has the following implications
- The "small K enablement" PR *revealed* the bug, but didn't actually cause it.
- The bug is in the "rewrite the case where the batch dimension's broadcast" handling we added a while back ... and backported to 6.1 because it was giving performance improvements despite the somewhat untested nature of that code
- Debugging will continue

**Comment by krzysz00** [comment]
This is now PR #1419, which'll need a release backport.
